content/shmem_team_split_3d.tex

```diff
@@ -1,22 +1,25 @@
 \apisummary{
-shmemx\_team\_split\_3d - partitions an existing parent team into three subgroups,
-based on the three-dimensional Cartesian space defined by the triplet (xrange,
-yrange, and zrange) describing the size of the Cartesian space in X, Y, and Z
-dimensions.
-}
+Create up to three new teams
+by splitting an existing parent team into up to three subsets based on a
+2D or 3D Cartesian space defined by the triplet (xrange, yrange, and zrange)
+describing the size of the Cartesian space in X, Y, and Z dimensions.}
 
 \begin{apidefinition}
 
 \begin{Csynopsis}
-void @\FuncDecl{shmemx\_team\_split\_3d}@(shmem_team_t parent_team, int xrange,
-int yrange, int zrange, shmem_team_t *xaxis_team, shmem_team_t *yaxis_team,
+void @\FuncDecl{shmem\_team\_split\_3d}@(shmem_team_t parent_team, long options,
+int xrange, int yrange, int zrange, shmem_team_t *xaxis_team, shmem_team_t *yaxis_team,
```

Comment:
Per the email I sent to the Teams Subcommittee, I think we should *only* specify `xrange` and `yrange` as parameters, otherwise a PE can call `split_3d` and not be in any of the resultant teams.

Reply:
This should probably be a response to the Teams-Subcommittee list, but I just sent out a request to join that list, so can't post there yet. I will post this same text there when I can.

It seems like that in the case where the programmer is going to be doing some work using the total number of processes in the current job in order to figure out their x, y, z dimension sizes, then they have already done the calculations to determine the best xrange, yrange for their job. So it is just a matter of whether to make the implementation take care of the zrange and to assign any trailing PEs into the final z coordinate.

In the email, the assumption is that if the program doesn't want to use the trailing PEs that don't fit nicely into the space, the programmer just ignores the final z plane, which will have some holes with no PEs.

So, in the case where the program has 72 PEs and the programmer has calculated that they want to have a 4x4x4 space, then if they use 3D split with x=4, y=4, the zrange will get assigned 5, so some of the PEs 0-64 will have a zteam of size 4, some a zteam of size 5.

But the 5th member of any zteam does not work on any of the 4x4x4 space data. And in this case, some of the PEs 0-64 would really want to broadcast, collect, barrier, etc with only the first 4 members of their 5 member zteam. The 5th wheel would just have to roll with it, har har. So this is fine if there will just be logic for some PEs to do no work and just join all the collectives. I could see this being better for library design since all the PEs stay in synch, even if they are just spinning.

Though, in the case of rather large machines and algorithms that work on regular grids, I could see some potential benefit for not including the unused PEs in any zteams teams to reduce unnecessary traffic. In that case, if someone wanted a 4x4x4 space with no unused PEs, they would do two team splits, first a strided split with size = 64, then a 3d split with xrange=4, yrange=4. But, since both splits would require synchronization overheads, it seems like it would be better if this case is covered by including the zrange param, since the only reason to reduce the number of processors in the first split is to avoid overheads in the teams resulting from the second split.

Does this sound right?

Reply:
> But the 5th member of any zteam does not work on any of the 4x4x4 space data. And in this case, some of the PEs 0-64 would really want to broadcast, collect, barrier, etc with only the first 4 members of their 5 member zteam.

There are probably a lot of assumptions on both sides here, but I think you're assuming that the 4x4x4 space you use for this example only has four meaningful elements of work in any dimension (i.e., the total space is partitioned into its smallest elements). There's nothing fundamentally that says the 5th PE has no meaningful work to do.

> [...] if someone wanted a 4x4x4 space with no unused PEs, they would do two team splits, first a strided split with size = 64, then a 3d split with xrange=4, yrange=4.

I think this is a good example showing the use of the `SHMEM_TEAM_NOCOLLECTIVE` option, which would be used on the initial strided split before the 3D split.

> But, since both splits would require synchronization overheads, it seems like it would be better if this case is covered by including the zrange param, since the only reason to reduce the number of processors in the first split is to avoid overheads in the teams resulting from the second split.

One high-level problem is that the 4x4x4 spatial decomposition and associated teams for a situation with 72 total PEs will generally result in some PEs doing no work (since they're not part of a team).

Reply:
One alternative that a colleague suggested is to **only** provide a 2D split with a single `xrange` parameter. The 3D split can be trivially derived from the 2D split by first splitting on the product of the desired *x*- and *y*-ranges.

One could make a similar argument to yours that a 2D split should provide extents for both *x*- and *y* dimensions, but I would argue (again?) that it's unintuitive to have all the PEs in a parent team call a 2D or 3D split and have some end up in no teams at all.

Reply:
I went ahead and started a new issue to change the 3D split function, so that this PR can just be for formatting and language fixes and we can start a new one for fixing the 3D split function.